pandas/core/groupby/generic.py

```diff
@@ -540,14 +540,23 @@ def _transform_general(
             if isinstance(res, (ABCDataFrame, ABCSeries)):
                 res = res._values
 
-            results.append(klass(res, index=group.index))
+            indexer = self._get_index(name) if self.dropna else group.index
+            results.append(klass(res, index=indexer))
 
         # check for empty "results" to avoid concat ValueError
         if results:
             from pandas.core.reshape.concat import concat
 
             concatenated = concat(results)
-            result = self._set_result_index_ordered(concatenated)
+
+            if self.dropna:
+                result = concatenated.sort_index()
+                if len(result.index) < len(self._selected_obj.index):
+                    result.index = self._selected_obj.index[result.index.asi8]
+                else:
+                    result.index = self._selected_obj.index
```

Comment:
this *could* go in _set_result_index_ordered not here. 
i am unclear why this is actually needed